File: golang/SlidingWindow/findAnagrams/findAnagrams.go
package findanagrams

func findanagrams(s string, p string) []int {
	sLen := len(s)
	pLen := len(p)

	if sLen < pLen {
		return nil
	}

	pCount := [26]int{}
	// windowCount := slices.Repeat([]int{0}, 26)
	windowCount := [26]int{}

	for i := 0; i < pLen; i++ {
		pCount[p[i]-'a']++
		windowCount[s[i]-'a']++
	}

	res := []int{}

	if pCount == windowCount {
		res = append(res, 0)
	}

	for right := pLen; right < sLen; right++ {
		left := right - pLen
		windowCount[s[left]-'a']--
		windowCount[s[right]-'a']++

		if windowCount == pCount {
			res = append(res, left+1)
		}
	}

	return res
}

File: Python/SlidingWindow/findAnagrams/findAnagrams.py
def findAnagrams(p: str, s: str) -> list[int]:
    sLen = len(s)
    pLen = len(p)

    res: list[int] = []

    if sLen < pLen:
        return res

    pCount = [0 for _ in range(26)]
    windowCount = [0 for _ in range(26)]

    for i in range(pLen):
        pCount[ord(p[i]) - ord("a")] += 1
        windowCount[ord(s[i]) - ord("a")] += 1

    if pCount == windowCount:
        res.append(0)

    for right in range(pLen, sLen):
        left = right - pLen
        windowCount[ord(s[left]) - ord("a")] -= 1
        windowCount[ord(s[right]) - ord("a")] += 1

        if windowCount == pCount:
            res.append(left + 1)

    return res
